### ISSUE TITLE
End to end tests

### ISSUE THREAD

**Issue by mikepurvis** [ros-drivers/rosserial#65]
Suggestion:

New rosserial_tests package, which contains a gtest executable and handful of rostest files. The gtest executable would be both a roscpp node and rosserial client in various configurations (pty, socket, etc).

The idea would be a baseline exercise of functionality, as well as potentially flood tests (for buffer overruns), correct handling of messages mangled in various ways, etc.

I'm happy to work on this.

**Comment by PaulBouchier** [comment]
I love it Mike! I've been thinking that we need exactly that - an automated
test suite so we can accept contributions and have confidence that they
don't break stuff. Contributions welcome.

Paul

**Comment by mikepurvis** [comment]
Yes, will be important to know about regressions, especially if there's serious effort put into tickets like #66 (just filed by me).

**Comment by PaulBouchier** [comment]
Here are some thoughts I've had on the topic:
- It might be desirable for them to run on the build farm, so that if dependencies change and break things they can be caught early.
- In any case, one would want to be able to run them on demand from a linux development box, without having to hook up some target embedded system and load firmware on it.
- ptys could be used on a Linux system to enable the server and client to communicate over a "serial link" when running an automated test. Sockets on localhost could be used for the same purpose. 
- A test harness would need to satisfy Arduino build dependencies. These are fairly minimal, as demonstrated by the EmbeddedLinux implementation, which does not depend on Arduino.

**Comment by mikepurvis** [comment]
Please check out my beginnings of some work on this:

https://github.com/clearpathrobotics/rosserial/tree/hydro-devel/rosserial_test

@mikeferguson, you may be interested in this too. There's some magic which happens in the client library generator script, and in the test fixture, but the meat of it is understandable from the simple check of publish functionality:

https://github.com/clearpathrobotics/rosserial/blob/hydro-devel/rosserial_test/src/publish_subscribe.cpp
